Enable deployment operations to utilize a custom ssh executable

tiwillia opened the pull request:
Bug 1177753
https://bugzilla.redhat.com/show_bug.cgi?id=1177753

Deploying a git ref, binary file, or binary file url previously only used ruby's implementation of ssh through the `Net::SSH` library. After this change, these deployment operations will honor the `--ssh` global option or `ssh` configuration directive.

Additionally, activating deployments will now also honor a custom ssh executable.

Still writing tests. All other tests are still passing, besides the missed test coverage.

tiwillia commented:
[test]

tiwillia commented:
Apparently I had accidently'd a few tests when moving them over to the `app_spec.rb`, [test] again please.

dobbymoodge commented on lib/rhc/commands/deployment.rb:
```diff
@@ -74,7 +74,9 @@ def activate(id)
-      activate_deployment(rest_app, id)
+      ssh_executable = check_ssh_executable! options.ssh
+
+      activate_deployment(rest_app, id, ssh_executable, !!options.ssh)
```
What's with `!!`?

tiwillia replied:
`!` in ruby (and in most dynamic languages) will return the opposite boolean value of the argument. In ruby, `nil` and `false` are the only things that evaluate to false. Since `options.ssh` will only ever either be `nil` or a string, `!!options.ssh` essentially is the same as `!options.ssh.nil?`, just shorter.

Miciah replied:
I think it would be cleaner to define `activate_deployment` with a single parameter instead of both `ssh_executable` and the `use_custom_ssh` Boolean and just pass in the value of `options.ssh && check_ssh_executable! options.ssh` as the argument here.  Is there a reason not to take that approach?

Miciah replied:
For that matter, maybe it would be cleaner to move the `options.ssh && check_ssh_executable! options.ssh` logic into `activate_deployment` itself (I think it has access to `options` as I describe in a comment below).  That would keep all the SSH-related logic together.

tiwillia commented:
re[test], net-ssh failures appear to be absolutely positively unrelated.

tiwillia commented:
Pull # 724 should resolve the test failures.

Miciah commented on lib/rhc/deployment_helpers.rb:
```diff
@@ -9,15 +9,17 @@ module DeploymentHelpers
 
     protected
 
-      def deploy_artifact(rest_app, artifact, hot_deploy, force_clean_build)
```
Because `deploy_artifact` is a method of `RHC::DeploymentHelpers`, which is used only as a mixin in `RHC::Commands::App` and `RHC::Commands::Deployment`, both of which subclass `RHC::Commands::Base`, which defines `options`, therefore the `options` parameter here is redundant (I think).

Miciah commented on lib/rhc/deployment_helpers.rb:
```diff
@@ -36,76 +38,104 @@ def deploy_artifact(rest_app, artifact, hot_deploy, force_clean_build)
         end
       end
 
+      def run_with_system_ssh(ssh_cmd)
```
Maybe this could be moved to `ssh_helpers.rb` and used in `RHC::Commands::PortForward` as well.

Miciah replied:
The `debug` statement would have to be factored out of `run_with_system_ssh` though—but it would be a net reduction in LOC.

Miciah commented on lib/rhc/deployment_helpers.rb:
```diff
@@ -36,76 +38,104 @@ def deploy_artifact(rest_app, artifact, hot_deploy, force_clean_build)
       end
 
-      def deploy_git_ref(rest_app, ref, hot_deploy, force_clean_build)
+      def deploy_git_ref(rest_app, ref, hot_deploy, force_clean_build, ssh_executable, use_custom_ssh)
```
I think `options` is available here (and also in `deploy_local_file` and `deploy_file_from_url`), so you could use `options.ssh` directly instead of the `use_custom_ssh` parameter.

Miciah commented on lib/rhc/deployment_helpers.rb:
```diff
@@ -9,15 +9,17 @@ module DeploymentHelpers
         is_file = File.file?(artifact)
         is_url = URI::ABS_URI.match(artifact).present?
 
+        ssh_executable = check_ssh_executable! options.ssh
```
I'd kind of rather push this down into `deploy_local_file`, `deploy_file_from_url`, and `deploy_git_ref` to avoid lengthening an already-long parameter list and to keep all the SSH-related logic together in those methods.

Miciah commented on spec/rhc/commands/app_spec.rb:
```diff
@@ -1002,4 +1002,342 @@
+      context "url file successfully" do
+        before do
+        @rest_app.stub(:deployment_type).and_return('binary')
+          ssh = double(Net::SSH)
```
What is the purpose of `ssh`? Did you mean to use it instead of `Net::SSH` below?

Miciah replied:
Same question applies to other tests where you have `double(Net:SSH)` and then don't use the double.

Miciah replied:
I just realised that this is code that you are moving from `deployment_spec.rb`, not code that you wrote, so never mind.

tiwillia replied:
:+1:

Miciah commented on spec/rhc/commands/app_spec.rb:
```diff
@@ -1002,4 +1002,342 @@
+    end
+
+    context "binary file successfully with ssh executable on windows" do
+      ssh_path = '/usr/bin/ssh'
```
Not a very Windowsy path. *grin*.  I guess it doesn't really matter though...

tiwillia commented:
@Miciah, thanks for the pointer.

I stopped passing options to the DeploymentHelpers methods, and am just using it there instead.

I also moved the `run_with_system_ssh` method to SshHelpers and changed the snapshot code to use that instead (which required modifying some tests).

Could you please re-review?

Miciah commented on lib/rhc/ssh_helpers.rb:
```diff
@@ -269,14 +269,10 @@ def save_snapshot(app, filename, for_deployment=false, ssh_executable=nil)
-            debug output
-            raise RHC::SnapshotSaveException.new "Error in trying to save snapshot. You can try to save manually by running:\n#{ssh_cmd}"
+            debug "Running #{ssh_cmd}"
+            status, output = run_with_system_ssh(ssh_cmd)
```
Nice clean-up!

tiwillia replied:
Thanks! :bow:

Miciah commented:
Please [merge]!
